## Load the CSV Files

In [1]:
# Add the Pandas dependency.
import pandas as pd

In [2]:
# Files to load
school_data_to_load = 'Resources\schools_complete.csv'
student_data_to_load = 'Resources\students_complete.csv'

In [3]:
# Note
# You can use indirect path method to access the files. You wil need to import the os module with your pandas dependency
# Add the dependencies.
#import pandas as pd
#import os
# Files to load
#school_data_to_load = os.path.join("Resources", "schools_complete.csv")
#student_data_to_load = os.path.join("Resources", "students_complete.csv")

## Read the School Data File

In [4]:
# Read the school data file and store it in a Pandas Dataframe.
school_data_df = pd.read_csv(school_data_to_load)
school_data_df

School ID            school_name      type  size   budget
0           0      Huang High School  District  2917  1910635
1           1   Figueroa High School  District  2949  1884411
2           2    Shelton High School   Charter  1761  1056600
3           3  Hernandez High School  District  4635  3022020
4           4    Griffin High School   Charter  1468   917500
5           5     Wilson High School   Charter  2283  1319574
6           6    Cabrera High School   Charter  1858  1081356
7           7     Bailey High School  District  4976  3124928
8           8     Holden High School   Charter   427   248087
9           9       Pena High School   Charter   962   585858
10         10     Wright High School   Charter  1800  1049400
11         11  Rodriguez High School  District  3999  2547363
12         12    Johnson High School  District  4761  3094650
13         13       Ford High School  District  2739  1763916
14         14     Thomas High School   Charter  1635  1043130

## Read the Student Data File

In [5]:
# Read the student data file and store it in a Pandas Dataframe.
student_data_df = pd.read_csv(student_data_to_load)
student_data_df.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

## Find Missing Values

### The count() Method -- counts the rows of each column containing data. 

In [6]:
# Determine if there are any missing values in the school data.
school_data_df.count()

School ID      15
school_name    15
type           15
size           15
budget         15
dtype: int64

In [7]:
# Determine if there are any missing values in the student data.
student_data_df.count()

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
dtype: int64

### The isnull() Method -- identifies rows with null values and returns boolean value: True for null and False for non-null value.

In [8]:
# Determine if there are any missing values in the school data.
school_data_df.isnull()

School ID  school_name   type   size  budget
0       False        False  False  False   False
1       False        False  False  False   False
2       False        False  False  False   False
3       False        False  False  False   False
4       False        False  False  False   False
5       False        False  False  False   False
6       False        False  False  False   False
7       False        False  False  False   False
8       False        False  False  False   False
9       False        False  False  False   False
10      False        False  False  False   False
11      False        False  False  False   False
12      False        False  False  False   False
13      False        False  False  False   False
14      False        False  False  False   False

In [9]:
# Determine if there are any missing values in the student data.
student_data_df.isnull()

Student ID  student_name  gender  grade  school_name  reading_score  \
0           False         False   False  False        False          False   
1           False         False   False  False        False          False   
2           False         False   False  False        False          False   
3           False         False   False  False        False          False   
4           False         False   False  False        False          False   
...           ...           ...     ...    ...          ...            ...   
39165       False         False   False  False        False          False   
39166       False         False   False  False        False          False   
39167       False         False   False  False        False          False   
39168       False         False   False  False        False          False   
39169       False         False   False  False        False          False   

       math_score  
0           False  
1           False  
2           False  
3           False  
4           False  
...           ...  
39165       False  
39166       False  
39167       False  
39168       False  
39169       False  

[39170 rows x 7 columns]

In [10]:
# To get the total number of empty rows, or rows that are "True," we can use the Pandas sum() method after the isnull() method, like this:
# Determine if there are any missing values in the student data.
student_data_df.isnull().sum()

Student ID       0
student_name     0
gender           0
grade            0
school_name      0
reading_score    0
math_score       0
dtype: int64

In [11]:
# Determine if there are any missing values in the school data.
school_data_df.isnull().sum()

School ID      0
school_name    0
type           0
size           0
budget         0
dtype: int64

### The notnull() Method -- identifies rows with non-null values and returns a boolean value: True for non-null and False for null value.

In [12]:
# Determine if there are not any missing values in the school data.
school_data_df.notnull()

School ID  school_name  type  size  budget
0        True         True  True  True    True
1        True         True  True  True    True
2        True         True  True  True    True
3        True         True  True  True    True
4        True         True  True  True    True
5        True         True  True  True    True
6        True         True  True  True    True
7        True         True  True  True    True
8        True         True  True  True    True
9        True         True  True  True    True
10       True         True  True  True    True
11       True         True  True  True    True
12       True         True  True  True    True
13       True         True  True  True    True
14       True         True  True  True    True

In [13]:
# Determine if there are not any missing values in the student data.
student_data_df.notnull().sum()

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
dtype: int64

## Determine data type

In [14]:
# Determine data types for the school DataFrame.
school_data_df.dtypes

School ID       int64
school_name    object
type           object
size            int64
budget          int64
dtype: object

### Note -- you can also determine the data type of a single column. If the column name does not have space, use df.column.dtype.
### But if the column name has space, use df['column'].dtype

In [15]:
school_data_df['School ID'].dtype

dtype('int64')

In [16]:
school_data_df.school_name.dtype

dtype('O')

In [17]:
# Determine data types for the student DataFrame.
student_data_df.dtypes

Student ID        int64
student_name     object
gender           object
grade            object
school_name      object
reading_score     int64
math_score        int64
dtype: object

In [18]:
# Add each prefix and suffix to remove to a list.
prefixes_suffixes = ["Dr. ", "Mr. ","Ms. ", "Mrs. ", "Miss ", " MD", " DDS", " DVM", " PhD"]

In [19]:
# Iterate through the words in the "prefixes_suffixes" list and replace them with an empty space, "".
for word in prefixes_suffixes:
    student_data_df["student_name"] = student_data_df["student_name"].str.replace(word,"")

C:\Users\Sa2di\.conda\envs\PythonEnv\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  This is separate from the ipykernel package so we can avoid doing imports until


In [20]:
student_data_df.head(10)

Student ID     student_name gender grade        school_name  reading_score  \
0           0     Paul Bradley      M   9th  Huang High School             66   
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
3           3    Richard Scott      M  12th  Huang High School             67   
4           4       Bonnie Ray      F   9th  Huang High School             97   
5           5    Bryan Miranda      M   9th  Huang High School             94   
6           6    Sheena Carter      F  11th  Huang High School             82   
7           7     Nicole Baker      F  12th  Huang High School             96   
8           8     Michael Roth      M  10th  Huang High School             95   
9           9   Matthew Greene      M  10th  Huang High School             96   

   math_score  
0          79  
1          61  
2          60  
3          58  
4          84  
5          94  
6          80  
7          69  
8          87  
9          84

## Merge DataFrames

The school district summary will be a high-level snapshot of the district's key metrics:

Total number of students. 
Total number of schools.
Total budget.
Average math score.
Average reading score.
Percentage of students who passed math.
Percentage of students who passed reading.
Overall passing percentage.

We'll merge school_data_df and student_data_dfon a shared column using the merge() method. The column that these DataFrames have in common is school_name. Inside the parentheses of the merge() method, we'll do the following:

Add the DataFrames to be merged.
Add the shared column to each DataFrame so that the merge can occur.
Define how the DataFrames should be merged: left, right, inner, or outer. The default is inner. (You will learn more about merging later in this course.)

In [21]:
# Combine teh data into a single dataset.
school_data_complete_df = pd.merge(student_data_df, school_data_df, on=['school_name', 'school_name'])
school_data_complete_df.head()

Student ID     student_name gender grade        school_name  reading_score  \
0           0     Paul Bradley      M   9th  Huang High School             66   
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
3           3    Richard Scott      M  12th  Huang High School             67   
4           4       Bonnie Ray      F   9th  Huang High School             97   

   math_score  School ID      type  size   budget  
0          79          0  District  2917  1910635  
1          61          0  District  2917  1910635  
2          60          0  District  2917  1910635  
3          58          0  District  2917  1910635  
4          84          0  District  2917  1910635

### Important
There may be cases in which you want to merge on a column that has similar information in two separate DataFrames, but is named differently in each—for example, "school_name" in one DataFrame and "high_school" in the second. In these cases, you should rename the columns so that they match. This will help avoid duplicate columns or merging issues.

## Get the Number of Students

Review
Recall that we counted the number of missing values in each DataFrame using the count() method. We can use the same method on the school_data_complete_df DataFrame to get the counts of specific columns.

In [22]:
# Get the total number of students.
student_count = school_data_complete_df.count()
student_count

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
School ID        39170
type             39170
size             39170
budget           39170
dtype: int64

In the output, we see that each column has 39,170 rows, or values. Therefore, we can choose any one of the columns and get the student count using the following format:
    school_data_complete_df[column].count()

In [23]:
# Get the student count.
student_count = school_data_complete_df["Student ID"].count()
student_count

39170

## Get the Number of Schools

### Reminder
remember we found teh total number of schools by using the count() method on the school_data_df dataFrame. The 
output was 15. 

There are two ways we can get the number of schools. One is to use the school_data_df DataFrame and assign a variable to one of the columns in the DataFrame, like this:

In [24]:
# Calculate the total number of schools.
school_count = school_data_df["school_name"].count()
school_count

15

We can't use the count() method on the school_data_complete_df["school_name"] column because this would give us a value of 39,170. If we want to use school_data_complete_df, we first need to get the unique items in the ["school_name"] column by using the unique() method. This method will return an array, or list, of all the unique values of that column.

In [25]:
# Calculate the total number of schools
school_count_2 = school_data_complete_df["school_name"].unique()
school_count_2

array(['Huang High School', 'Figueroa High School', 'Shelton High School',
       'Hernandez High School', 'Griffin High School',
       'Wilson High School', 'Cabrera High School', 'Bailey High School',
       'Holden High School', 'Pena High School', 'Wright High School',
       'Rodriguez High School', 'Johnson High School', 'Ford High School',
       'Thomas High School'], dtype=object)

In [26]:
len(school_count_2)

15

## Get the Total Budget

In [27]:
for school_name in school_count_2:
    Total_budget = school_data_complete_df["budget"].unique().sum()
print(Total_budget)    

24649428


### Alternatively

In [28]:
# Calculate the total budget.
total_budget = school_data_df["budget"].sum()
total_budget

24649428

## Get the Score Averages

The Pandas method for getting the average of columns is the mean() method.

### Find the Average Reading Score

In [29]:
# Calculate the average reading score.
average_reading_score = school_data_complete_df['reading_score'].mean()
average_reading_score

81.87784018381414

In [30]:
# Format the result to 2 decimal place
print(f'The average reading score is {average_reading_score:.2f}%')

The average reading score is 81.88%


### Find the Average Math Score

In [31]:
# Calculate the average math score.
average_math_score = school_data_complete_df["math_score"].mean()
average_math_score

78.98537145774827

In [32]:
# Format the result to 2 decimal place
print(f'The average math score is {average_math_score:.2f}%')

The average math score is 78.99%


## Get the Passing Percentages

To get the percentage of students who passed math and the percentage of students who passed reading, we will write code to:

Determine the passing grade.
Get the number of students who passed math and the number of students who passed reading in separate DataFrames.
Calculate the number of students who passed math and the number of students who passed reading.
Calculate the percentage of students who passed math and the percentage of students who passed reading.

To get the overall passing percentage, we will write code to:

Get the number of students who passed both math and reading in a DataFrame.
Calculate the number of students who passed both math and reading.
Calculate the percentage of students who passed both math and reading.

### Determine the Passing Grade
For math and reading assessment tests in this school district, the passing score was 70. Therefore, we need to get all the math and reading scores that are greater than or equal to 70. To do this, in a new cell, assign a passing_math variable to the math_score column in school_data_complete_df, where all the math scores are equal to or greater than 70.

In [33]:
passing_math = school_data_complete_df['math_score'] >= 70
passing_math.head()

0     True
1    False
2    False
3    False
4     True
Name: math_score, dtype: bool

In [34]:
passing_reading = school_data_complete_df['reading_score'] >= 70
passing_reading.head()

0    False
1     True
2     True
3    False
4     True
Name: reading_score, dtype: bool

### Get the Number of Students Who Passed Math and Reading
To get all the students who passed math and all the students who passed reading, we need to filter our school_data_complete_df DataFrame for the "True" cases. In other words, get only the students who have a grade is equal or greater to 70.

We can filter the school_data_complete_df DataFrame by adding the school_data_complete_df["math_score"] >= 70 within brackets, like this:

In [35]:
# Get all the students who are passing math in a new DataFrame.
passing_math = school_data_complete_df[school_data_complete_df["math_score"] >= 70]
passing_math.head()

Student ID   student_name gender grade        school_name  reading_score  \
0           0   Paul Bradley      M   9th  Huang High School             66   
4           4     Bonnie Ray      F   9th  Huang High School             97   
5           5  Bryan Miranda      M   9th  Huang High School             94   
6           6  Sheena Carter      F  11th  Huang High School             82   
8           8   Michael Roth      M  10th  Huang High School             95   

   math_score  School ID      type  size   budget  
0          79          0  District  2917  1910635  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635  
8          87          0  District  2917  1910635

In [36]:
# Get all the students that are passing reading in a new DataFrame.
passing_reading = school_data_complete_df[school_data_complete_df["reading_score"] >= 70]
passing_reading.head()

Student ID     student_name gender grade        school_name  reading_score  \
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
4           4       Bonnie Ray      F   9th  Huang High School             97   
5           5    Bryan Miranda      M   9th  Huang High School             94   
6           6    Sheena Carter      F  11th  Huang High School             82   

   math_score  School ID      type  size   budget  
1          61          0  District  2917  1910635  
2          60          0  District  2917  1910635  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635

In [37]:
# Calculate the number of students passing math.
passing_math_count = passing_math['student_name'].count()
passing_math_count

29370

In [38]:
# Calculate the number of students passing reading.
passing_reading_count = passing_reading['student_name'].count()
passing_reading_count

33610

## Get the percentage of students who passed math and reading.
To get the percentage of students who passed math and reading, divide the passing_math_count and the passing_reading_count by the total number of students, and then multiply by 100.

Reminder

Recall that the student count was calculated using this code:
    school_data_complete_df['Student ID'].count() and it gave us 39,170

Since we are calculating a percentage, we need to convert the student_count to a number with a decimal, or floating-point decimal, by using float().

The final calculation should look like this:

In [39]:
# Calculate the percentage that pased math.
passing_math_percentage = passing_math_count / float(student_count) * 100
print(passing_math_percentage)

# Calculate the percentage that pased reading.
passing_reading_percentage = passing_reading_count / float(student_count) * 100
print(passing_reading_percentage)

74.9808526933878
85.80546336482001


## Calculate the overall passing percentage
To get the overall passing percentage, we need to get all the students who passed both math and reading and divide by the total number of students.

We can filter the school_data_complete_df DataFrame by adding the school_data_complete_df["math_score"] >= 70 and school_data_complete_df["reading_score"] >= 70 with the logical operator "&" within brackets, like this:

In [40]:
# Calculate the students who passed both math and reading.
passing_math_reading = school_data_complete_df[(school_data_complete_df["math_score"] >= 70) & (school_data_complete_df["reading_score"] >= 70)]

passing_math_reading.head()

Student ID    student_name gender grade        school_name  reading_score  \
4           4      Bonnie Ray      F   9th  Huang High School             97   
5           5   Bryan Miranda      M   9th  Huang High School             94   
6           6   Sheena Carter      F  11th  Huang High School             82   
8           8    Michael Roth      M  10th  Huang High School             95   
9           9  Matthew Greene      M  10th  Huang High School             96   

   math_score  School ID      type  size   budget  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635  
8          87          0  District  2917  1910635  
9          84          0  District  2917  1910635

In [41]:
# Calculate the number of students who passed both math and reading.
overall_passing_math_reading_count = passing_math_reading["student_name"].count()
overall_passing_math_reading_count

25528

Finally, we calculate the percentage of students who passed both math and reading by dividing the total number of students and multiplying by 100, using the following code.

In [42]:
# Calculate the overall passing percentage.
overall_passing_percentage = overall_passing_math_reading_count / student_count * 100
overall_passing_percentage

65.17232575950983

## Create a District Summary DataFrame
Now that we have performed all the calculations needed for the district summary, let's add the following values and columns to a new DataFrame named district_summary_df.

Total number of schools in the column "Total Schools"
Total number of students in the column "Total Students"
Total budget in the column "Total Budget"
Average reading score in the column "Average Reading Score"
Average math score in the column "Average Math Score"
Percentage of students passing reading in the column "% Passing Reading"
Percentage of students passing math in the column "% Passing Math"
Overall passing percentage in the column "% Overall Passing"

Remember, one way to create a new DataFrame is to convert a list of dictionaries to a DataFrame.

In [43]:
# Adding a list of values with keys to create a new DataFrame.
district_summary_df = pd.DataFrame(
          [{"Total Schools": school_count,
          "Total Students": student_count,
          "Total Budget": total_budget,
          "Average Math Score": average_math_score,
          "Average Reading Score": average_reading_score,
          "% Passing Math": passing_math_percentage,
         "% Passing Reading": passing_reading_percentage,
        "% Overall Passing": overall_passing_percentage}])
district_summary_df

Total Schools  Total Students  Total Budget  Average Math Score  \
0             15           39170      24649428           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  % Overall Passing  
0               81.87784       74.980853          85.805463          65.172326

## Format Columns

To clean up the district_summary_df DataFrame, we will format dollar amounts to two decimal places, and format the grade averages to one decimal place and percentages to the nearest whole number percent.

##### NOTE
Grade averages are usually formatted to one decimal place because averages taken to the hundredths (range: 0.01–0.09) or lower do not impact an average as much as tenths of a grade point, (range: 0.1–0.9). When tenths of a grade point equal or are greater than 0.5, you should usually round up to the next whole number.

Grade percentages are formatted to the whole percentage because the tenths of a percent is equivalent to thousandths of a value, or grade. Like averages, thousandths of a grade don't have an impact on the overall percentage because they are so small.

This type of formatting can be done with the built-in Pandas map() function. The map() function is used for substituting each value in a Series with another value. Where the new value is generated from a function, a dictionary, or a Series.

### Writing Functions for the School District Data

In [44]:
# define a function that calculates the percentage of students that passed 
# both math and reading and print the passing percentage to the output when the 
# function is called.
def passing_math_percentage(passing_math_count, student_count):
    return passing_math_count / float(student_count) * 100

In [45]:
# call the function
passing_math_percentage(passing_math_count, student_count)

74.9808526933878

Now that you have a good handle on how to write and use functions, let's apply this knowledge to the map() function, which we'll use to format our data.

### The Pandas map() function

Let's apply this formatting technique to format the Total Students column with a thousands separator using the basic syntax:

df["column"] = df["column"].map("{:,}".format)

In [46]:
# Format the "Total Students" to have the comma for a thousands separator.
district_summary_df["Total Students"] = district_summary_df["Total Students"].map("{:,}".format)

district_summary_df["Total Students"]

0    39,170
Name: Total Students, dtype: object

In [47]:
# Format "Total Budget" to have the comma for a thousands separator, a decimal separator, and a "$".

district_summary_df["Total Budget"] = district_summary_df["Total Budget"].map("${:,.2f}".format)

district_summary_df["Total Budget"]

0    $24,649,428.00
Name: Total Budget, dtype: object

We need to format the remaining columns accordingly:

The "Average Reading Score" column will be formatted to one decimal place.
The "Average Math Score" column will be formatted to one decimal place.
The "% Passing Reading" column will be formatted to the nearest whole number percentage.
The "% Passing Math" column will be formatted to the nearest whole number percentage
The "% Overall Passing" column will be formatted to the nearest whole number percentage.

In [48]:
# Format the columns.
district_summary_df["Average Math Score"] = district_summary_df["Average Math Score"].map("{:.1f}".format)

district_summary_df["Average Reading Score"] = district_summary_df["Average Reading Score"].map("{:.1f}".format)

district_summary_df["% Passing Math"] = district_summary_df["% Passing Math"].map("{:.0f}".format)

district_summary_df["% Passing Reading"] = district_summary_df["% Passing Reading"].map("{:.0f}".format)

district_summary_df["% Overall Passing"] = district_summary_df["% Overall Passing"].map("{:.0f}".format)

In [49]:
# display the dataframe.
district_summary_df

Total Schools Total Students    Total Budget Average Math Score  \
0             15         39,170  $24,649,428.00               79.0   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0                  81.9             75                86                65

## Reorder Columns

To reorder columns using Pandas, we can pass a list of columns to a current DataFrame using square bracket notation. This tells Pandas to select those specific columns and put them in the DataFrame in the same order that they appear in the list.

Here is the standard format for reordering columns:

### Reorder the columns in the order you want them to appear.
new_column_order = ["column2", "column4", "column1"]

### Assign a new or the same DataFrame the new column order.
df = df[new_column_order]
This technique can also be used to filter out columns that you don't need and select the columns in the order that you want. For example, we didn't include column3 in the new order; this is how we can filter out columns we don't want.

To get the columns in the correct order, add the following code and run the cell.

### Reorder the columns in the order you want them to appear.
new_column_order = ["Total Schools", "Total Students", "Total Budget","Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]

### Assign district summary df the new column order.
district_summary_df = district_summary_df[new_column_order]
district_summary_df
The DataFrame in the output cell should look like this:

In [50]:
# Reorder the columns in the order you want them to appear.
new_column_order = ["Total Schools", "Total Students", "Total Budget","Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]

# Assign district summary df the new column order.
district_summary_df = district_summary_df[new_column_order]
district_summary_df

Total Schools Total Students    Total Budget Average Math Score  \
0             15         39,170  $24,649,428.00               79.0   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0                  81.9             75                86                65

## Set the Index to the School Name

In [51]:
# determine the school type.
per_school_types = school_data_df.set_index(['school_name'])['type']
per_school_types

school_name
Huang High School        District
Figueroa High School     District
Shelton High School       Charter
Hernandez High School    District
Griffin High School       Charter
Wilson High School        Charter
Cabrera High School       Charter
Bailey High School       District
Holden High School        Charter
Pena High School          Charter
Wright High School        Charter
Rodriguez High School    District
Johnson High School      District
Ford High School         District
Thomas High School        Charter
Name: type, dtype: object

Now we'll create a new DataFrame by converting this Series to a DataFrame as follows:

In [52]:
# Add the per_school_types into a DataFrame for testing.
df = pd.DataFrame(per_school_types)
df

type
school_name                    
Huang High School      District
Figueroa High School   District
Shelton High School     Charter
Hernandez High School  District
Griffin High School     Charter
Wilson High School      Charter
Cabrera High School     Charter
Bailey High School     District
Holden High School      Charter
Pena High School        Charter
Wright High School      Charter
Rodriguez High School  District
Johnson High School    District
Ford High School       District
Thomas High School      Charter

## IMPORTANT
As we get the rest of the data for this DataFrame, we need to make sure the index for the data being added as columns is always "school_name."

## Get the Student Count Per School

Looking at the school_data_df DataFrame, we can see the student count is in the "size" column.

Now let's get the student count in the school_data_df DataFrame. Add the following code to a new cell and run the cell.

When we run this cell, we get the following output: a Series showing the number of students in each school with a numerical index (0–14), which is the same as the index of the school_data_df DataFrame.

In [53]:
# calculate the total student count.
per_school_counts = school_data_df['size']
per_school_counts

0     2917
1     2949
2     1761
3     4635
4     1468
5     2283
6     1858
7     4976
8      427
9      962
10    1800
11    3999
12    4761
13    2739
14    1635
Name: size, dtype: int64

Unfortunately, this Series doesn't have an index with "school_name." Therefore, we can't use the "size" column from school_data_df to get the count of the student population.

To fix this, we can use the set_index() method on the "school_name" column, and then select the "size" column to display the student count for each school.

In [54]:
# Calculate the total student count.
per_school_counts = school_data_df.set_index(["school_name"])["size"]
per_school_counts

school_name
Huang High School        2917
Figueroa High School     2949
Shelton High School      1761
Hernandez High School    4635
Griffin High School      1468
Wilson High School       2283
Cabrera High School      1858
Bailey High School       4976
Holden High School        427
Pena High School          962
Wright High School       1800
Rodriguez High School    3999
Johnson High School      4761
Ford High School         2739
Thomas High School       1635
Name: size, dtype: int64

Next, we'll use the school_data_complete_df DataFrame

To get the number of students from the school_data_complete_df DataFrame, we can count the number of times a high school appears using value_counts() on the "school_name" column. The value_counts() method will return a Series of data with the number of times each school_name appears in a row. 

In [55]:
# Calculate the total student count.
per_school_complete_counts = school_data_complete_df["school_name"].value_counts()
per_school_complete_counts

Bailey High School       4976
Johnson High School      4761
Hernandez High School    4635
Rodriguez High School    3999
Figueroa High School     2949
Huang High School        2917
Ford High School         2739
Wilson High School       2283
Cabrera High School      1858
Wright High School       1800
Shelton High School      1761
Thomas High School       1635
Griffin High School      1468
Pena High School          962
Holden High School        427
Name: school_name, dtype: int64

## Get the Budget Per Student

In order to find the budget per student for each school, we [need to divide the budget for each school by the number of students at each school.

We can use the set_index() method on the school_name column of the school_data_df DataFrame to get the school_name as the index, like we did when we calculated the total students from each school, with school_data_df.set_index(["school_name"])["size"].

In [56]:
# calculate the total school budget.
per_school_budget = school_data_df.set_index(["school_name"])["budget"]
per_school_budget

school_name
Huang High School        1910635
Figueroa High School     1884411
Shelton High School      1056600
Hernandez High School    3022020
Griffin High School       917500
Wilson High School       1319574
Cabrera High School      1081356
Bailey High School       3124928
Holden High School        248087
Pena High School          585858
Wright High School       1049400
Rodriguez High School    2547363
Johnson High School      3094650
Ford High School         1763916
Thomas High School       1043130
Name: budget, dtype: int64

The data type for the budget column is int64, which is suitable for calculations that we need to perform in order to find the budget per student.

To get the budget per student, we'll divide the per_school_budget by the per_school_counts.

In [57]:
# Calculate the per capita spending.
per_school_capita = per_school_budget / per_school_counts
per_school_capita

school_name
Huang High School        655.0
Figueroa High School     639.0
Shelton High School      600.0
Hernandez High School    652.0
Griffin High School      625.0
Wilson High School       578.0
Cabrera High School      582.0
Bailey High School       628.0
Holden High School       581.0
Pena High School         609.0
Wright High School       583.0
Rodriguez High School    637.0
Johnson High School      650.0
Ford High School         644.0
Thomas High School       638.0
dtype: float64

We can perform this calculation because the per_school_budget and per_school_counts are Series, both data types are int64, and both have the same index.

## Get the Score Averages Per School

We've used the set_index() method on the school_name column in student_data_df to get data from another column, just like how we retrieved the school budget using school_data_df.set_index(["school_name"])["budget"].

Let's use this procedure to replace budget with math_score

In [58]:
# Calculate the math scores.
student_school_math = student_data_df.set_index(["school_name"])["math_score"]
student_school_math

school_name
Huang High School     79
Huang High School     61
Huang High School     60
Huang High School     58
Huang High School     84
                      ..
Thomas High School    90
Thomas High School    70
Thomas High School    84
Thomas High School    90
Thomas High School    75
Name: math_score, Length: 39170, dtype: int64

Unfortunately, we can't use the school_data_df DataFrame, as there aren't any columns containing grades. We also can't use the set_index() method on the school_name column in student_data_df because there are too many occurrences of the school_name column.

Instead, we need to use the Pandas groupby() function. The groupby() function will split an object (like a DataFrame), apply a mathematical operation, and combine the results. This can be used to group large amounts of data when we want to compute mathematical operations on these groups.

In [59]:
# Calculate the average math scores.
per_school_averages = school_data_complete_df.groupby(["school_name"]).mean()
per_school_averages

Student ID  reading_score  math_score  School ID  \
school_name                                                               
Bailey High School        20358.5      81.033963   77.048432        7.0   
Cabrera High School       16941.5      83.975780   83.061895        6.0   
Figueroa High School       4391.0      81.158020   76.711767        1.0   
Ford High School          36165.0      80.746258   77.102592       13.0   
Griffin High School       12995.5      83.816757   83.351499        4.0   
Hernandez High School      9944.0      80.934412   77.289752        3.0   
Holden High School        23060.0      83.814988   83.803279        8.0   
Huang High School          1458.0      81.182722   76.629414        0.0   
Johnson High School       32415.0      80.966394   77.072464       12.0   
Pena High School          23754.5      84.044699   83.839917        9.0   
Rodriguez High School     28035.0      80.744686   76.842711       11.0   
Shelton High School        6746.0      83.725724   83.359455        2.0   
Thomas High School        38352.0      83.848930   83.418349       14.0   
Wilson High School        14871.0      83.989488   83.274201        5.0   
Wright High School        25135.5      83.955000   83.682222       10.0   

                         size     budget  
school_name                               
Bailey High School     4976.0  3124928.0  
Cabrera High School    1858.0  1081356.0  
Figueroa High School   2949.0  1884411.0  
Ford High School       2739.0  1763916.0  
Griffin High School    1468.0   917500.0  
Hernandez High School  4635.0  3022020.0  
Holden High School      427.0   248087.0  
Huang High School      2917.0  1910635.0  
Johnson High School    4761.0  3094650.0  
Pena High School        962.0   585858.0  
Rodriguez High School  3999.0  2547363.0  
Shelton High School    1761.0  1056600.0  
Thomas High School     1635.0  1043130.0  
Wilson High School     2283.0  1319574.0  
Wright High School     1800.0  1049400.0

## Calculate the average test scores.

In [60]:
per_school_math = per_school_averages['math_score']
per_school_math 

school_name
Bailey High School       77.048432
Cabrera High School      83.061895
Figueroa High School     76.711767
Ford High School         77.102592
Griffin High School      83.351499
Hernandez High School    77.289752
Holden High School       83.803279
Huang High School        76.629414
Johnson High School      77.072464
Pena High School         83.839917
Rodriguez High School    76.842711
Shelton High School      83.359455
Thomas High School       83.418349
Wilson High School       83.274201
Wright High School       83.682222
Name: math_score, dtype: float64

In [61]:
per_school_reading = per_school_averages['reading_score']
per_school_reading

school_name
Bailey High School       81.033963
Cabrera High School      83.975780
Figueroa High School     81.158020
Ford High School         80.746258
Griffin High School      83.816757
Hernandez High School    80.934412
Holden High School       83.814988
Huang High School        81.182722
Johnson High School      80.966394
Pena High School         84.044699
Rodriguez High School    80.744686
Shelton High School      83.725724
Thomas High School       83.848930
Wilson High School       83.989488
Wright High School       83.955000
Name: reading_score, dtype: float64

## Get the passing percentages per school.

 To get the passing percentages, we need to:
 1. Determine what is the passing grade.
 2. Get the number of students who passed math and reading.
 3. Get the students who passed math and passed reading

#### Determine the Passing Grade

Remember, for assessment tests, the passing score is 70, with the >= 70 statement to filter the grades that are passing.

#### Get the Number of Students Who Passed Math and Reading

We determined the number of students who passed math and reading for the district summary using the following code:

passing_math = school_data_complete_df[(school_data_complete_df["math_score"] >= 70)]

and

passing_reading = school_data_complete_df[(school_data_complete_df["reading_score"] >= 70)]

We'll use the same code but assign each calculation to new variables that reflect the students who passed math or reading for each school

In [62]:
# Calculate the passing scores by creating a filtered DataFrame.
per_school_passing_math = school_data_complete_df[(school_data_complete_df["math_score"] >= 70)]
per_school_passing_math.head()

Student ID   student_name gender grade        school_name  reading_score  \
0           0   Paul Bradley      M   9th  Huang High School             66   
4           4     Bonnie Ray      F   9th  Huang High School             97   
5           5  Bryan Miranda      M   9th  Huang High School             94   
6           6  Sheena Carter      F  11th  Huang High School             82   
8           8   Michael Roth      M  10th  Huang High School             95   

   math_score  School ID      type  size   budget  
0          79          0  District  2917  1910635  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635  
8          87          0  District  2917  1910635

In [63]:
per_school_passing_reading = school_data_complete_df[(school_data_complete_df["reading_score"] >= 70)]
per_school_passing_reading.head()

Student ID     student_name gender grade        school_name  reading_score  \
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
4           4       Bonnie Ray      F   9th  Huang High School             97   
5           5    Bryan Miranda      M   9th  Huang High School             94   
6           6    Sheena Carter      F  11th  Huang High School             82   

   math_score  School ID      type  size   budget  
1          61          0  District  2917  1910635  
2          60          0  District  2917  1910635  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635

To create a Series that has school_name as the index, we'll use groupby(["school_name"]) on the per_school_passing_math and the per_school_passing_math DataFrames.

In [64]:
# Calculate the number of students passing math and passing reading by school.
per_school_passing_math = per_school_passing_math.groupby(["school_name"]).count()["student_name"]
per_school_passing_math

school_name
Bailey High School       3318
Cabrera High School      1749
Figueroa High School     1946
Ford High School         1871
Griffin High School      1371
Hernandez High School    3094
Holden High School        395
Huang High School        1916
Johnson High School      3145
Pena High School          910
Rodriguez High School    2654
Shelton High School      1653
Thomas High School       1525
Wilson High School       2143
Wright High School       1680
Name: student_name, dtype: int64

In [65]:
per_school_passing_reading = per_school_passing_reading.groupby(["school_name"]).count()["student_name"]
per_school_passing_reading

school_name
Bailey High School       4077
Cabrera High School      1803
Figueroa High School     2381
Ford High School         2172
Griffin High School      1426
Hernandez High School    3748
Holden High School        411
Huang High School        2372
Johnson High School      3867
Pena High School          923
Rodriguez High School    3208
Shelton High School      1688
Thomas High School       1591
Wilson High School       2204
Wright High School       1739
Name: student_name, dtype: int64

#### Determine the Percentage of Students Passing Math and Reading
To determine the percentage of students passing math and reading, we must divide per_school_passing_math and per_school_passing_reading by the per_school_counts, and then multiply by 100, as shown in the following code:

In [66]:
# Calculate the percentage of passing math and reading scores per school.
per_school_passing_math = per_school_passing_math / per_school_counts * 100
per_school_passing_math

school_name
Bailey High School       66.680064
Cabrera High School      94.133477
Figueroa High School     65.988471
Ford High School         68.309602
Griffin High School      93.392371
Hernandez High School    66.752967
Holden High School       92.505855
Huang High School        65.683922
Johnson High School      66.057551
Pena High School         94.594595
Rodriguez High School    66.366592
Shelton High School      93.867121
Thomas High School       93.272171
Wilson High School       93.867718
Wright High School       93.333333
dtype: float64

In [67]:
per_school_passing_reading = per_school_passing_reading / per_school_counts * 100
per_school_passing_reading

school_name
Bailey High School       81.933280
Cabrera High School      97.039828
Figueroa High School     80.739234
Ford High School         79.299014
Griffin High School      97.138965
Hernandez High School    80.862999
Holden High School       96.252927
Huang High School        81.316421
Johnson High School      81.222432
Pena High School         95.945946
Rodriguez High School    80.220055
Shelton High School      95.854628
Thomas High School       97.308869
Wilson High School       96.539641
Wright High School       96.611111
dtype: float64

## Get the Overall Passing Percentage for All Students for Each School

To get the overall passing percentage, we need to get all the students who passed both math and reading and then divide by the total number of students.

In [68]:
# Calculate the students who passed both math and reading.
per_passing_math_reading = school_data_complete_df[(school_data_complete_df["math_score"] >= 70) & (school_data_complete_df["reading_score"] >= 70)]

per_passing_math_reading.head()

Student ID    student_name gender grade        school_name  reading_score  \
4           4      Bonnie Ray      F   9th  Huang High School             97   
5           5   Bryan Miranda      M   9th  Huang High School             94   
6           6   Sheena Carter      F  11th  Huang High School             82   
8           8    Michael Roth      M  10th  Huang High School             95   
9           9  Matthew Greene      M  10th  Huang High School             96   

   math_score  School ID      type  size   budget  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635  
8          87          0  District  2917  1910635  
9          84          0  District  2917  1910635

To get the total number of students who passed both math and reading on the per_passing_math_reading DataFrame, we use the following code, which sets the index to school_name, and then use the count() method for the student_name.

In [69]:
# Calculate the number of students who passed both math and reading.
per_passing_math_reading = per_passing_math_reading.groupby(["school_name"]).count()["student_name"]
per_passing_math_reading

school_name
Bailey High School       2719
Cabrera High School      1697
Figueroa High School     1569
Ford High School         1487
Griffin High School      1330
Hernandez High School    2481
Holden High School        381
Huang High School        1561
Johnson High School      2549
Pena High School          871
Rodriguez High School    2119
Shelton High School      1583
Thomas High School       1487
Wilson High School       2068
Wright High School       1626
Name: student_name, dtype: int64

Finally, we calculate the percentage of students who passed math and reading by dividing the the total number of students and multiplying by 100, using the following code.

In [70]:
# Calculate the overall passing percentage.
per_overall_passing_percentage = per_passing_math_reading / per_school_counts * 100
per_overall_passing_percentage

school_name
Bailey High School       54.642283
Cabrera High School      91.334769
Figueroa High School     53.204476
Ford High School         54.289887
Griffin High School      90.599455
Hernandez High School    53.527508
Holden High School       89.227166
Huang High School        53.513884
Johnson High School      53.539172
Pena High School         90.540541
Rodriguez High School    52.988247
Shelton High School      89.892107
Thomas High School       90.948012
Wilson High School       90.582567
Wright High School       90.333333
dtype: float64

# Create the School Summary DataFrame

Now that we have performed all the calculations needed for the school summary, we can add the following values to a new DataFrame named per_school_summary_df. The data and columns of the DataFrame will be:

Type of school in the "School Type" column
Total students per school in the "Total Students" column
Total budget per school in the "Total School Budget" column
Total budget per student for each school in the "Per Student Budget" column
Average math score for each school in the "Average Math Score" column
Average reading score for each school in the "Average Reading Score" column
Percentage of students passing math for each school in the "% Passing Math" column
Percentage of students passing reading for each school in the "% Passing Reading" column
Overall passing percentage for each school in the "% Overall Passing" column

In [71]:
# Adding a list of values with keys to create a new DataFrame.

per_school_summary_df = pd.DataFrame({
             "School Type": per_school_types,
             "Total Students": per_school_counts,
             "Total School Budget": per_school_budget,
             "Per Student Budget": per_school_capita,
             "Average Math Score": per_school_math,
           "Average Reading Score": per_school_reading,
           "% Passing Math": per_school_passing_math,
           "% Passing Reading": per_school_passing_reading,
           "% Overall Passing": per_overall_passing_percentage})
per_school_summary_df.head()

School Type  Total Students  Total School Budget  \
school_name                                                             
Bailey High School      District            4976              3124928   
Cabrera High School      Charter            1858              1081356   
Figueroa High School    District            2949              1884411   
Ford High School        District            2739              1763916   
Griffin High School      Charter            1468               917500   

                      Per Student Budget  Average Math Score  \
school_name                                                    
Bailey High School                 628.0           77.048432   
Cabrera High School                582.0           83.061895   
Figueroa High School               639.0           76.711767   
Ford High School                   644.0           77.102592   
Griffin High School                625.0           83.351499   

                      Average Reading Score  % Passing Math  \
school_name                                                   
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  
school_name                                                 
Bailey High School            81.933280          54.642283  
Cabrera High School           97.039828          91.334769  
Figueroa High School          80.739234          53.204476  
Ford High School              79.299014          54.289887  
Griffin High School           97.138965          90.599455

# Clean Up the DataFrame

In the per_school_summary_df DataFrame, we will format the Total Students, Total School Budget, Per Student Budget, Average Math Score, Average Reading Score, % Passing Math. % Passing Reading, and % Overall Passing columns to include:

A U.S. dollar sign 

Two decimal places (optional)

One decimal places

Zero decimal places

A thousands separator

## Format columns

In [72]:
# Format the "Total Students" to have the comma for a thousands separator.
per_school_summary_df["Total Students"] = per_school_summary_df["Total Students"].map("{:,}".format)

per_school_summary_df["Total Students"]

school_name
Bailey High School       4,976
Cabrera High School      1,858
Figueroa High School     2,949
Ford High School         2,739
Griffin High School      1,468
Hernandez High School    4,635
Holden High School         427
Huang High School        2,917
Johnson High School      4,761
Pena High School           962
Rodriguez High School    3,999
Shelton High School      1,761
Thomas High School       1,635
Wilson High School       2,283
Wright High School       1,800
Name: Total Students, dtype: object

In [73]:
# Format the "Total School Budget" to have the comma for a thousands separator and dollar sign, $.
per_school_summary_df["Total School Budget"] = per_school_summary_df["Total School Budget"].map("${:,}".format)

per_school_summary_df["Total School Budget"]

school_name
Bailey High School       $3,124,928
Cabrera High School      $1,081,356
Figueroa High School     $1,884,411
Ford High School         $1,763,916
Griffin High School        $917,500
Hernandez High School    $3,022,020
Holden High School         $248,087
Huang High School        $1,910,635
Johnson High School      $3,094,650
Pena High School           $585,858
Rodriguez High School    $2,547,363
Shelton High School      $1,056,600
Thomas High School       $1,043,130
Wilson High School       $1,319,574
Wright High School       $1,049,400
Name: Total School Budget, dtype: object

In [74]:
# Format the "Per Student Budget" to have the comma for a thousands separator and dollar sign, $.
per_school_summary_df["Per Student Budget"] = per_school_summary_df["Per Student Budget"].map("${:,}".format)

per_school_summary_df["Per Student Budget"]

school_name
Bailey High School       $628.0
Cabrera High School      $582.0
Figueroa High School     $639.0
Ford High School         $644.0
Griffin High School      $625.0
Hernandez High School    $652.0
Holden High School       $581.0
Huang High School        $655.0
Johnson High School      $650.0
Pena High School         $609.0
Rodriguez High School    $637.0
Shelton High School      $600.0
Thomas High School       $638.0
Wilson High School       $578.0
Wright High School       $583.0
Name: Per Student Budget, dtype: object

In [75]:
# Format the columns.
per_school_summary_df["Average Math Score"] = per_school_summary_df["Average Math Score"].map("{:.1f}".format)

per_school_summary_df["Average Reading Score"] = per_school_summary_df["Average Reading Score"].map("{:.1f}".format)

per_school_summary_df["% Passing Math"] = per_school_summary_df["% Passing Math"].map("{:.0f}".format)

per_school_summary_df["% Passing Reading"] = per_school_summary_df["% Passing Reading"].map("{:.0f}".format)

per_school_summary_df["% Overall Passing"] = per_school_summary_df["% Overall Passing"].map("{:.0f}".format)

In [76]:
per_school_summary_df.head()

School Type Total Students Total School Budget  \
school_name                                                           
Bailey High School      District          4,976          $3,124,928   
Cabrera High School      Charter          1,858          $1,081,356   
Figueroa High School    District          2,949          $1,884,411   
Ford High School        District          2,739          $1,763,916   
Griffin High School      Charter          1,468            $917,500   

                     Per Student Budget Average Math Score  \
school_name                                                  
Bailey High School               $628.0               77.0   
Cabrera High School              $582.0               83.1   
Figueroa High School             $639.0               76.7   
Ford High School                 $644.0               77.1   
Griffin High School              $625.0               83.4   

                     Average Reading Score % Passing Math % Passing Reading  \
school_name                                                                   
Bailey High School                    81.0             67                82   
Cabrera High School                   84.0             94                97   
Figueroa High School                  81.2             66                81   
Ford High School                      80.7             68                79   
Griffin High School                   83.8             93                97   

                     % Overall Passing  
school_name                             
Bailey High School                  55  
Cabrera High School                 91  
Figueroa High School                53  
Ford High School                    54  
Griffin High School                 91

# Find the Highest-Performing Schools

to find the highest performing schools, you sort the datafrmae above based on the % overall passing using sort_value().

In addition, we can add the parameter ascending=False to sort from highest to lowest, or ascending=True to sort from lowest to highest the value referenced in the sort_values() function. If we don't add the ascending parameter, the default is ascending=True

In [77]:
# Sort and show top five schools.
top_schools = per_school_summary_df.sort_values(["% Overall Passing"], ascending=False)

top_schools.head(10)

School Type Total Students Total School Budget  \
school_name                                                          
Cabrera High School     Charter          1,858          $1,081,356   
Griffin High School     Charter          1,468            $917,500   
Pena High School        Charter            962            $585,858   
Thomas High School      Charter          1,635          $1,043,130   
Wilson High School      Charter          2,283          $1,319,574   
Shelton High School     Charter          1,761          $1,056,600   
Wright High School      Charter          1,800          $1,049,400   
Holden High School      Charter            427            $248,087   
Bailey High School     District          4,976          $3,124,928   
Ford High School       District          2,739          $1,763,916   

                    Per Student Budget Average Math Score  \
school_name                                                 
Cabrera High School             $582.0               83.1   
Griffin High School             $625.0               83.4   
Pena High School                $609.0               83.8   
Thomas High School              $638.0               83.4   
Wilson High School              $578.0               83.3   
Shelton High School             $600.0               83.4   
Wright High School              $583.0               83.7   
Holden High School              $581.0               83.8   
Bailey High School              $628.0               77.0   
Ford High School                $644.0               77.1   

                    Average Reading Score % Passing Math % Passing Reading  \
school_name                                                                  
Cabrera High School                  84.0             94                97   
Griffin High School                  83.8             93                97   
Pena High School                     84.0             95                96   
Thomas High School                   83.8             93                97   
Wilson High School                   84.0             94                97   
Shelton High School                  83.7             94                96   
Wright High School                   84.0             93                97   
Holden High School                   83.8             93                96   
Bailey High School                   81.0             67                82   
Ford High School                     80.7             68                79   

                    % Overall Passing  
school_name                            
Cabrera High School                91  
Griffin High School                91  
Pena High School                   91  
Thomas High School                 91  
Wilson High School                 91  
Shelton High School                90  
Wright High School                 90  
Holden High School                 89  
Bailey High School                 55  
Ford High School                   54

# Find the Lowest-Performing Schools

o sort for the five lowest-performing schools based on the % Overall Passing

In [78]:
# Sort and show top five schools.
bottom_schools = per_school_summary_df.sort_values(["% Overall Passing"], ascending=True)

bottom_schools.head()

School Type Total Students Total School Budget  \
school_name                                                            
Figueroa High School     District          2,949          $1,884,411   
Rodriguez High School    District          3,999          $2,547,363   
Ford High School         District          2,739          $1,763,916   
Hernandez High School    District          4,635          $3,022,020   
Huang High School        District          2,917          $1,910,635   

                      Per Student Budget Average Math Score  \
school_name                                                   
Figueroa High School              $639.0               76.7   
Rodriguez High School             $637.0               76.8   
Ford High School                  $644.0               77.1   
Hernandez High School             $652.0               77.3   
Huang High School                 $655.0               76.6   

                      Average Reading Score % Passing Math % Passing Reading  \
school_name                                                                    
Figueroa High School                   81.2             66                81   
Rodriguez High School                  80.7             66                80   
Ford High School                       80.7             68                79   
Hernandez High School                  80.9             67                81   
Huang High School                      81.2             66                81   

                      % Overall Passing  
school_name                              
Figueroa High School                 53  
Rodriguez High School                53  
Ford High School                     54  
Hernandez High School                54  
Huang High School                    54

# Create Grade-Level DataFrames

You are tasked with creating two new DataFrames: one that will display the average math scores for each grade, and the one that will display the average reading scores for each grade. These will be grouped by the name of the school.

In [79]:
# create a grade level dataframes.
ninth_graders = school_data_complete_df[(school_data_complete_df['grade'] == '9th')]
tenth_graders = school_data_complete_df[(school_data_complete_df['grade'] == '10th')]
eleventh_graders = school_data_complete_df[(school_data_complete_df['grade'] == '11th')]
twelfth_graders = school_data_complete_df[(school_data_complete_df['grade'] == '12th')]

In [80]:
ninth_graders.head()

Student ID     student_name gender grade        school_name  \
0            0     Paul Bradley      M   9th  Huang High School   
4            4       Bonnie Ray      F   9th  Huang High School   
5            5    Bryan Miranda      M   9th  Huang High School   
12          12  Brittney Walker      F   9th  Huang High School   
13          13     William Long      M   9th  Huang High School   

    reading_score  math_score  School ID      type  size   budget  
0              66          79          0  District  2917  1910635  
4              97          84          0  District  2917  1910635  
5              94          94          0  District  2917  1910635  
12             64          79          0  District  2917  1910635  
13             71          79          0  District  2917  1910635

In [81]:
tenth_graders.head()

Student ID      student_name gender grade        school_name  \
8            8      Michael Roth      M  10th  Huang High School   
9            9    Matthew Greene      M  10th  Huang High School   
10          10  Andrew Alexander      M  10th  Huang High School   
11          11     Daniel Cooper      M  10th  Huang High School   
14          14      Tammy Hebert      F  10th  Huang High School   

    reading_score  math_score  School ID      type  size   budget  
8              95          87          0  District  2917  1910635  
9              96          84          0  District  2917  1910635  
10             90          70          0  District  2917  1910635  
11             78          77          0  District  2917  1910635  
14             85          67          0  District  2917  1910635

# Score Averages Grouped by School Name

## Get the Average Math Scores by School
First, we will get the average math score for each school. For each grade level DataFrame, we will use the groupby() function on the school_name column and apply the mean() on the math_score column

In [82]:
# Group each grade level DataFrame by the school name for the average math score.
ninth_grade_math_scores = ninth_graders.groupby(["school_name"]).mean()["math_score"]

tenth_grade_math_scores = tenth_graders.groupby(["school_name"]).mean()["math_score"]

eleventh_grade_math_scores = eleventh_graders.groupby(["school_name"]).mean()["math_score"]

twelfth_grade_math_scores = twelfth_graders.groupby(["school_name"]).mean()["math_score"]

In [83]:
ninth_grade_math_scores

school_name
Bailey High School       77.083676
Cabrera High School      83.094697
Figueroa High School     76.403037
Ford High School         77.361345
Griffin High School      82.044010
Hernandez High School    77.438495
Holden High School       83.787402
Huang High School        77.027251
Johnson High School      77.187857
Pena High School         83.625455
Rodriguez High School    76.859966
Shelton High School      83.420755
Thomas High School       83.590022
Wilson High School       83.085578
Wright High School       83.264706
Name: math_score, dtype: float64

In [84]:
eleventh_grade_math_scores

school_name
Bailey High School       77.515588
Cabrera High School      82.765560
Figueroa High School     76.884344
Ford High School         76.918058
Griffin High School      83.842105
Hernandez High School    77.136029
Holden High School       85.000000
Huang High School        76.446602
Johnson High School      77.491653
Pena High School         84.328125
Rodriguez High School    76.395626
Shelton High School      83.383495
Thomas High School       83.498795
Wilson High School       83.195326
Wright High School       83.836782
Name: math_score, dtype: float64

## Get the Average Reading Scores by School
First, we will get the average reading score for each school. For each grade level DataFrame, we will use the groupby() function on the school_name column and apply the mean() on the reading_score column.

In [85]:
# Group each grade level DataFrame by the school name for the average math score.
ninth_grade_reading_scores = ninth_graders.groupby(["school_name"]).mean()["reading_score"]

tenth_grade_reading_scores = tenth_graders.groupby(["school_name"]).mean()["reading_score"]

eleventh_grade_reading_scores = eleventh_graders.groupby(["school_name"]).mean()["reading_score"]

twelfth_grade_reading_scores = twelfth_graders.groupby(["school_name"]).mean()["reading_score"]

In [86]:
twelfth_grade_reading_scores

school_name
Bailey High School       80.912451
Cabrera High School      84.287958
Figueroa High School     81.384863
Ford High School         80.662338
Griffin High School      84.013699
Hernandez High School    80.857143
Holden High School       84.698795
Huang High School        80.305983
Johnson High School      81.227564
Pena High School         84.591160
Rodriguez High School    80.376426
Shelton High School      82.781671
Thomas High School       83.831361
Wilson High School       84.317673
Wright High School       84.073171
Name: reading_score, dtype: float64

In [87]:
twelfth_grade_reading_scores['Shelton High School']

82.78167115902966

# Combine each grade level Series into a DataFrame

In [88]:
# Combine each grade level Series for average math scores by school into a single DataFrame.
math_scores_by_grade = pd.DataFrame({
               "9th": ninth_grade_math_scores,
               "10th": tenth_grade_math_scores,
               "11th": eleventh_grade_math_scores,
               "12th": twelfth_grade_math_scores})

math_scores_by_grade.head()

9th       10th       11th       12th
school_name                                                     
Bailey High School    77.083676  76.996772  77.515588  76.492218
Cabrera High School   83.094697  83.154506  82.765560  83.277487
Figueroa High School  76.403037  76.539974  76.884344  77.151369
Ford High School      77.361345  77.672316  76.918058  76.179963
Griffin High School   82.044010  84.229064  83.842105  83.356164

In [89]:
# Combine each grade level Series for average reading scores by school into a single DataFrame.
reading_scores_by_grade = pd.DataFrame({
               "9th": ninth_grade_reading_scores,
               "10th": tenth_grade_reading_scores,
               "11th": eleventh_grade_reading_scores,
               "12th": twelfth_grade_reading_scores})

reading_scores_by_grade.head()

9th       10th       11th       12th
school_name                                                     
Bailey High School    81.303155  80.907183  80.945643  80.912451
Cabrera High School   83.676136  84.253219  83.788382  84.287958
Figueroa High School  81.198598  81.408912  80.640339  81.384863
Ford High School      80.632653  81.262712  80.403642  80.662338
Griffin High School   83.369193  83.706897  84.288089  84.013699

# Format the Averages and Remove the Index Name

You will format the grade-level averages to one decimal place, as well as remove the name of the index column, school_name

In [90]:
# format the math scores by grade
math_scores_by_grade['9th'] = math_scores_by_grade['9th'].map('{:.1f}'.format)
math_scores_by_grade['10th'] = math_scores_by_grade['10th'].map('{:.1f}'.format)
math_scores_by_grade['11th'] = math_scores_by_grade['11th'].map('{:.1f}'.format)
math_scores_by_grade['12th'] = math_scores_by_grade['12th'].map('{:.1f}'.format)

In [91]:
# Format the reading scores by grade
reading_scores_by_grade['9th'] = reading_scores_by_grade['9th'].map('{:.1f}'.format)
reading_scores_by_grade['10th'] = reading_scores_by_grade['10th'].map('{:.1f}'.format)
reading_scores_by_grade['11th'] = reading_scores_by_grade['11th'].map('{:.1f}'.format)
reading_scores_by_grade['12th'] = reading_scores_by_grade['12th'].map('{:.1f}'.format)

In [92]:
math_scores_by_grade.head()

9th  10th  11th  12th
school_name                                 
Bailey High School    77.1  77.0  77.5  76.5
Cabrera High School   83.1  83.2  82.8  83.3
Figueroa High School  76.4  76.5  76.9  77.2
Ford High School      77.4  77.7  76.9  76.2
Griffin High School   82.0  84.2  83.8  83.4

In [93]:
reading_scores_by_grade.head()

9th  10th  11th  12th
school_name                                 
Bailey High School    81.3  80.9  80.9  80.9
Cabrera High School   83.7  84.3  83.8  84.3
Figueroa High School  81.2  81.4  80.6  81.4
Ford High School      80.6  81.3  80.4  80.7
Griffin High School   83.4  83.7  84.3  84.0

In [94]:
# Remove the index name, school_name
math_scores_by_grade.index.name = None

# Display the DataFrame.
math_scores_by_grade.head()

9th  10th  11th  12th
Bailey High School    77.1  77.0  77.5  76.5
Cabrera High School   83.1  83.2  82.8  83.3
Figueroa High School  76.4  76.5  76.9  77.2
Ford High School      77.4  77.7  76.9  76.2
Griffin High School   82.0  84.2  83.8  83.4

In [95]:
# Remove the index name, school_name
reading_scores_by_grade.index.name = None

# Display the DataFrame.
reading_scores_by_grade.head()

9th  10th  11th  12th
Bailey High School    81.3  80.9  80.9  80.9
Cabrera High School   83.7  84.3  83.8  84.3
Figueroa High School  81.2  81.4  80.6  81.4
Ford High School      80.6  81.3  80.4  80.7
Griffin High School   83.4  83.7  84.3  84.0

# Establish the spending ranges per student

## Get the spending ranges
Before we create the bins, we need to determine the distribution of spending per student. We can find the distribution for school spending per student by using the describe() method.

When we apply the describe() method to a DataFrame or Series, it will return the following descriptive statistics for the DataFrame or Series:

The number of rows in the DataFrame or Series as count
The average of the rows as mean
The standard deviation of the rows as std
The minimum value of the rows as min
The 25th percentile as 25%
The 50th percentile as 50%
The 75th percentile as 75%
The minimum value of the rows as max

In [96]:
# Get the descriptive statistics for the per_school_capita.
per_school_capita.describe()

count     15.000000
mean     620.066667
std       28.544368
min      578.000000
25%      591.500000
50%      628.000000
75%      641.500000
max      655.000000
dtype: float64

As you can see, the minimum is 578 and the maximum is 655. The standard deviation is 28.5, or approximately 30. We don't want the lowest bin to be $578 because there is only one school at or below $578, which is Wilson High School. However, there are four schools that spend less than $585 per student, so $585 will be our lowest bin. Also, because the standard deviation is about 30, we will increase the bins by $30, up to $675. Therefore, the four bins will be: $585, $615, $645, and $675.

In [97]:
# In pandas, we can write the ranges for the bins as follows:
spending_bins = [0, 585, 615, 645, 675]

## Group the Series on the Spending Ranges

To group the spending ranges, we use the Pandas cut() function. The cut() function segments and sorts data values into bins. We'll use the cut() function to create our spending bins.

In [98]:
# Cut the per_school_capita into the spending ranges.
spending_bins = [0, 585, 615, 645, 675]
pd.cut(per_school_capita, spending_bins)

school_name
Huang High School        (645, 675]
Figueroa High School     (615, 645]
Shelton High School      (585, 615]
Hernandez High School    (645, 675]
Griffin High School      (615, 645]
Wilson High School         (0, 585]
Cabrera High School        (0, 585]
Bailey High School       (615, 645]
Holden High School         (0, 585]
Pena High School         (585, 615]
Wright High School         (0, 585]
Rodriguez High School    (615, 645]
Johnson High School      (645, 675]
Ford High School         (615, 645]
Thomas High School       (615, 645]
dtype: category
Categories (4, interval[int64, right]): [(0, 585] < (585, 615] < (615, 645] < (645, 675]]

Next, we'll determine how many schools are in each range by grouping the spending bins as the index using the groupby() function on the per_school_capita Series. Remember that we need to perform a mathematical operation when we use the groupby() function.

To find out how many schools are in those bins, use the count() method on the groupby() function.

In [99]:
# Cut the per_school_capita into the spending ranges.
spending_bins = [0, 585, 615, 645, 675]
per_school_capita.groupby(pd.cut(per_school_capita, spending_bins)).count()

(0, 585]      4
(585, 615]    2
(615, 645]    6
(645, 675]    3
dtype: int64

Looking at this output, it seems that we didn't group the schools fairly. We need to adjust our ranges so that we have three or four schools in each range. Let's increase the [585, 615] range to [585, 630] and change the third range to [630, 645]. The spending_bins will look like this: 

[0, 585, 630, 645, 675]

In [100]:
# Cut the per_school_capita into the spending ranges.
spending_bins = [0, 585, 630, 645, 675]
per_school_capita.groupby(pd.cut(per_school_capita, spending_bins)).count()

(0, 585]      4
(585, 630]    4
(630, 645]    4
(645, 675]    3
dtype: int64

You can label the ranges using a list of string values.

In [101]:
# Establish the spending bins and group names.
spending_bins = [0, 585, 630, 645, 675]
group_names = ["<$584", "$585-629", "$630-644", "$645-675"]

# Categorize the spending bins

In [102]:
# Categorize spending based on the bins.
per_school_summary_df["Spending Ranges (Per Student)"] = pd.cut(per_school_capita, spending_bins, labels=group_names)

per_school_summary_df.head()

School Type Total Students Total School Budget  \
school_name                                                           
Bailey High School      District          4,976          $3,124,928   
Cabrera High School      Charter          1,858          $1,081,356   
Figueroa High School    District          2,949          $1,884,411   
Ford High School        District          2,739          $1,763,916   
Griffin High School      Charter          1,468            $917,500   

                     Per Student Budget Average Math Score  \
school_name                                                  
Bailey High School               $628.0               77.0   
Cabrera High School              $582.0               83.1   
Figueroa High School             $639.0               76.7   
Ford High School                 $644.0               77.1   
Griffin High School              $625.0               83.4   

                     Average Reading Score % Passing Math % Passing Reading  \
school_name                                                                   
Bailey High School                    81.0             67                82   
Cabrera High School                   84.0             94                97   
Figueroa High School                  81.2             66                81   
Ford High School                      80.7             68                79   
Griffin High School                   83.8             93                97   

                     % Overall Passing Spending Ranges (Per Student)  
school_name                                                           
Bailey High School                  55                      $585-629  
Cabrera High School                 91                         <$584  
Figueroa High School                53                      $630-644  
Ford High School                    54                      $630-644  
Griffin High School                 91                      $585-629

In [103]:
# Calculate averages for the desired columns.

spending_math_scores = per_school_summary_df.groupby(["Spending Ranges (Per Student)", "Average Math Score"]).mean()[["Average Math Score"]]

spending_reading_scores = per_school_summary_df.groupby(["Spending Ranges (Per Student)", "Average Reading Score"]).mean()[["Average Reading Score"]]

spending_passing_math = per_school_summary_df.groupby(["Spending Ranges (Per Student)", "% Passing Math"]).mean()[["% Passing Math"]]

spending_passing_reading = per_school_summary_df.groupby(["Spending Ranges (Per Student)", "% Passing Reading"]).mean()[["% Passing Reading"]]

overall_passing_spending = per_school_summary_df.groupby(["Spending Ranges (Per Student)", "% Overall Passing"]).mean()[["% Overall Passing"]]

C:\Users\Sa2di\.conda\envs\PythonEnv\lib\site-packages\IPython\core\interactiveshell.py:3444: FutureWarning: Dropping invalid columns in DataFrameGroupBy.mean is deprecated. In a future version, a TypeError will be raised. Before calling .mean, select only columns which should be valid for the function.
  exec(code_obj, self.user_global_ns, self.user_ns)


KeyError: "None of [Index(['Average Math Score'], dtype='object')] are in the [columns]"

In [ ]:
spending_math_scores

In [ ]:
# spending_reading_scores

In [ ]:
# spending_passing_math

# Create the spending summary dataframe

In [ ]:
# Assemble into dataframe.
spending_summary_df = pd.DataFrame({"Average Math Score" : spending_math_scores,
          "Average Reading Score": spending_reading_scores,
          "% Passing Math": spending_passing_math,
          "% Passing Reading": spending_passing_reading,
          "% Overall Passing": overall_passing_spending})
spending_summary_df